### PR TITLE
Update `datasets` to work with integer timestamp

### PR DESCRIPTION
**Pull request by d-a-bunin**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Closing issues
#139.

### CODE REVIEW

File: etna/datasets/tsdataset.py [hunk 1]
```diff
@@ -278,27 +313,20 @@ def make_future(
...
-
-        new_index = self.raw_df.index.append(future_dates)
-        df = self.raw_df.reindex(new_index)
-        df.index.name = "timestamp"
+        df = self._expand_index(df=self.raw_df, freq=self.freq, future_steps=future_steps)
 
         if self.df_exog is not None and self.current_df_level == self.current_df_exog_level:
-            df = self._merge_exog(df)
+            # current implementation doesn't fail here if it doesn't fail in __init__, so check_regressors is redundant
+            df = self._merge_exog(df=df, check_regressors=False)
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:320]: I'm not sure we should touch this here, I could remove this. 

But we probably don't need this check here, it doen't fail here if it doesn't fail in `__init__`.

File: etna/datasets/tsdataset.py [hunk 2]
```diff
@@ -278,27 +313,20 @@ def make_future(
...
 
             # check if we have enough values in regressors
             if self.regressors:
-                for segment in self.segments:
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:295]: I simplified this check, I should also remove it from this PR.

File: etna/datasets/tsdataset.py [hunk 3]
```diff
@@ -574,8 +610,12 @@ def plot(
             k = len(segments)
         columns_num = min(2, k)
         rows_num = math.ceil(k / columns_num)
-        start = self.df.index.min() if start is None else pd.Timestamp(start)
-        end = self.df.index.max() if end is None else pd.Timestamp(end)
+
+        start = _check_timestamp_param(param=start, param_name="start", freq=self.freq)
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:614]: It seems like I forgot to check this in tests.

**Comment by ostreech1997** [comment]
We usually create `TSDataset` like this:
```python
TSDataset(TSDataset.to_dataset(df), freq=freq)
```
So, I think, we should add here additional info about possible types of "timestamp" column in docstring:
https://github.com/etna-team/etna/blob/f1990a96989f5675fb09df172ae2591cc569ec40/etna/datasets/tsdataset.py#L766

File: etna/datasets/tsdataset.py [hunk 4]
```diff
@@ -203,13 +192,43 @@ def fit_transform(self, transforms: Sequence["Transform"]):
...
+    def _prepare_df(cls, df: pd.DataFrame, freq: Optional[str]) -> pd.DataFrame:
+        # cast segment to str type
+        cls._cast_segment_to_str(df)
+
+        # handle freq
+        if freq is None:
+            if df.index.dtype != "int":
+                raise ValueError("You set wrong freq. Data contains datetime index, not integer.")
+
+            new_index = np.arange(df.index.min(), df.index.max())
```
Comment (ostreech1997) [inline etna/datasets/tsdataset.py:211]: Isn't that how it should be
```python
new_index = np.arange(df.index.min(), df.index.max() + 1)
```

File: etna/datasets/tsdataset.py (hunk 1 above)
Comment (ostreech1997) [reply etna/datasets/tsdataset.py:320]: It seems that you are right.

File: etna/datasets/tsdataset.py [hunk 5]
```diff
@@ -230,6 +249,22 @@ def __getitem__(self, item):
...
 
+    @staticmethod
+    def _expand_index(df: pd.DataFrame, freq: Optional[str], future_steps: int) -> pd.DataFrame:
+        if freq is None:
+            new_index = np.arange(df.index.min(), df.index.max() + future_steps + 1)
+        else:
+            with warnings.catch_warnings():
+                warnings.filterwarnings(action="ignore", message="Argument `closed` is deprecated")
+                future_dates = pd.date_range(start=df.index.max(), periods=future_steps + 1, freq=freq, closed="right")
+            new_index = df.index.append(future_dates)
```
Comment (ostreech1997) [inline etna/datasets/tsdataset.py:260]: Why do we generate `new_index` for int "timestamp" from `min` to `max + future_steps +1`, when for date "timestamp" we generate from `max` to `max + future_steps + 1` and then append to the original index? Is there a hidden logic here?

File: etna/datasets/tsdataset.py (hunk 2 above)
Comment (ostreech1997) [reply etna/datasets/tsdataset.py:295]: Do you mean that it will be better to do in separate issue?

File: tests/test_datasets/test_dataset.py [hunk 6]
```diff
@@ -310,6 +322,93 @@ def ts_with_prediction_intervals(ts_without_target_components, prediction_interv
     return ts
 
 
+def test_create_ts_with_datetime_timestamp():
+    freq = "D"
+    df = generate_ar_df(periods=10, freq=freq, n_segments=3)
+    df_wide = TSDataset.to_dataset(df)
+    df_wide.index.freq = freq
```
Comment (ostreech1997) [inline tests/test_datasets/test_dataset.py:329]: Why we do it?

File: tests/test_datasets/test_dataset.py [hunk 7]
```diff
@@ -310,6 +322,93 @@ def ts_with_prediction_intervals(ts_without_target_components, prediction_interv
...
+    freq = "D"
+    df = generate_ar_df(periods=10, start_time="2020-01-05", freq=freq, n_segments=3, random_seed=0)
+    df_exog = generate_ar_df(periods=20, start_time="2020-01-01", freq=freq, n_segments=3, random_seed=1)
+    df_exog.rename(columns={"target": "exog"}, inplace=True)
+
+    df_wide = TSDataset.to_dataset(df)
+    df_exog_wide = TSDataset.to_dataset(df_exog)
+    ts = TSDataset(df=df_wide, df_exog=df_exog_wide, freq=freq)
+
+    expected_merged = pd.concat([df_wide, df_exog_wide.loc[df_wide.index]], axis=1).sort_index(axis=1)
```
Comment (ostreech1997) [inline tests/test_datasets/test_dataset.py:355]: In `TSDataset._merge_exog` we use `sort_index` with `level=(0, 1)`. May be here we also should use it?
```python
expected_merged = pd.concat([df_wide, df_exog_wide.loc[df_wide.index]], axis=1).sort_index(axis=1, level=(0, 1))
```

File: tests/test_datasets/test_dataset.py [hunk 8]
```diff
@@ -310,6 +322,93 @@ def ts_with_prediction_intervals(ts_without_target_components, prediction_interv
...
+def test_create_ts_with_exog_int_timestamp():
+    df = generate_ar_df(periods=10, start_time=5, freq=None, n_segments=3, random_seed=0)
+    df_exog = generate_ar_df(periods=20, start_time=0, freq=None, n_segments=3, random_seed=1)
+    df_exog.rename(columns={"target": "exog"}, inplace=True)
+
+    df_wide = TSDataset.to_dataset(df)
+    df_exog_wide = TSDataset.to_dataset(df_exog)
+    ts = TSDataset(df=df_wide, df_exog=df_exog_wide, freq=None)
+
+    expected_merged = pd.concat([df_wide, df_exog_wide.loc[df_wide.index]], axis=1).sort_index(axis=1)
```
Comment (ostreech1997) [inline tests/test_datasets/test_dataset.py:370]: Same as comment above.

File: etna/datasets/tsdataset.py (hunk 2 above)
Comment (d-a-bunin) [reply etna/datasets/tsdataset.py:295]: Yes, I think it could be done in a separate issue.

File: etna/datasets/tsdataset.py (hunk 5 above)
Comment (d-a-bunin) [reply etna/datasets/tsdataset.py:260]: For int timestamp it is easy to make `arange`.

For datetime timestamp I just keep current code.

**Comment by ostreech1997** [comment]
In the [issue ](https://github.com/etna-team/etna/issues/139)  there is two points:
- Clarify in the docs that timestamp could be integer for `describe` method
- Clarify in the docs that timestamp could be integer for `info` method

Should we take them into account in this task?

**Review by ostreech1997 — APPROVED** [review]

**Pull request merged**